File: libsql/typescript/src/adapter.ts
import { createClient, type Client, type Row } from "@libsql/client";
import type {
  DatabaseAdapter,
  QueryOptions,
  QueryResult,
  TableSummary,
  TableDescription,
  ColumnInfo,
  IndexInfo,
  ForeignKeyInfo,
  Secret,
} from "@database-mcp/core";

/** BIGINTs beyond Number.MAX_SAFE_INTEGER are returned as strings (JSON-safe). */
function jsonSafe(value: unknown): unknown {
  if (typeof value === "bigint") {
    return value >= Number.MIN_SAFE_INTEGER && value <= Number.MAX_SAFE_INTEGER
      ? Number(value)
      : value.toString();
  }
  return value;
}

/**
 * libSQL is a SQLite fork: catalog queries are identical to the SQLite
 * adapter's. What differs is the transport — @libsql/client speaks to local
 * files (file:) and remote servers (libsql:/https:) with an auth token.
 */
export class LibsqlAdapter implements DatabaseAdapter {
  engine = "libsql";
  private client!: Client;

  constructor(
    private url: string,
    private authToken?: Secret,
  ) {}

  async connect({ readOnly }: { readOnly: boolean }): Promise<void> {
    // A plain filesystem path is accepted for parity with @database-mcp/sqlite.
    const url = /^[a-z][a-z0-9+.-]*:/i.test(this.url) ? this.url : `file:${this.url}`;
    this.client = createClient({
      url,
      authToken: this.authToken?.reveal(),
      intMode: "bigint",
    });
    if (readOnly) {
      try {
        await this.client.execute("PRAGMA query_only = ON");
      } catch {
        // Remote servers may not honor per-session pragmas. The SQL guard
        // still blocks writes; for real protection use a read-only token.
        console.error("libsql: session read-only unavailable on this server — use a read-only auth token");
      }
    }
    // Fail fast on bad URL/credentials.
    await this.client.execute("SELECT 1");
  }

  async close(): Promise<void> {
    this.client.close();
  }

  async query(sql: string, { maxRows, timeoutMs }: QueryOptions): Promise<QueryResult> {
    // ponytail: deadline via Promise.race — the HTTP request itself is not
    // cancelled, only abandoned; per-request AbortSignal if it ever matters.
    const timeout = new Promise<never>((_, reject) =>
      setTimeout(() => reject(new Error(`timeout: query exceeded ${timeoutMs}ms`)), timeoutMs).unref(),
    );
    const rs = await Promise.race([this.client.execute(sql), timeout]);

    if (rs.columns.length === 0) {
      return { columns: [], rows: [], rowCount: Number(rs.rowsAffected), truncated: false };
    }
    const truncated = rs.rows.length > maxRows;
    const rows = rs.rows.slice(0, maxRows).map((row: Row) => {
      const out: Record<string, unknown> = {};
      for (const col of rs.columns) out[col] = jsonSafe(row[col]);
      return out;
    });
    return { columns: [...rs.columns], rows, rowCount: rows.length, truncated };
  }

  async listTables(): Promise<TableSummary[]> {
    const tables = await this.client.execute(
      "SELECT name FROM sqlite_master WHERE type = 'table' AND name NOT LIKE 'sqlite_%' ORDER BY name",
    );
    const out: TableSummary[] = [];
    for (const row of tables.rows) {
      const name = String(row.name);
      const count = await this.client.execute(`SELECT count(*) AS n FROM "${name.replace(/"/g, '""')}"`);
      out.push({ name, estimatedRows: Number(count.rows[0].n) });
    }
    return out;
  }

  async describeTable(table: string): Promise<TableDescription> {
    const cols = await this.client.execute({ sql: "SELECT * FROM pragma_table_info(?)", args: [table] });
    if (cols.rows.length === 0) throw new Error(`unknown table: ${table}`);

    const columns: ColumnInfo[] = cols.rows.map((c) => ({
      name: String(c.name),
      type: String(c.type),
      nullable: !Number(c.notnull),
      key: Number(c.pk) ? "PRI" : null,
      default: c.dflt_value === null ? null : String(c.dflt_value),
    }));

    const indexList = await this.client.execute({ sql: "SELECT * FROM pragma_index_list(?)", args: [table] });
    const indexes: IndexInfo[] = [];
    for (const ix of indexList.rows) {
      const info = await this.client.execute({ sql: "SELECT name FROM pragma_index_info(?)", args: [String(ix.name)] });
      indexes.push({
        name: String(ix.name),
        columns: info.rows.map((c) => String(c.name)),
        unique: !!Number(ix.unique),
      });
    }

    const fkRows = await this.client.execute({ sql: "SELECT * FROM pragma_foreign_key_list(?)", args: [table] });
    const fkById = new Map<string, ForeignKeyInfo>();
    for (const fk of fkRows.rows) {
      const key = String(fk.id);
      const entry = fkById.get(key) ?? {
        name: `fk_${table}_${key}`, // SQLite-family FKs are unnamed; synthesize a stable one
        columns: [],
        referencesTable: String(fk.table),
        referencesColumns: [],
      };
      entry.columns.push(String(fk.from));
      entry.referencesColumns.push(String(fk.to));
      fkById.set(key, entry);
    }

    return { name: table, columns, indexes, foreignKeys: [...fkById.values()] };
  }
}

File: libsql/python/src/database_mcp_libsql/adapter.py
"""libSQL adapter over the libsql driver. Zero MCP types.

libSQL is a SQLite fork: catalog queries are identical to the SQLite
adapter's. What differs is the transport: the driver speaks to local files
and remote servers (libsql:/https:/ws:) with an auth token.
"""

import concurrent.futures
import sys
from typing import Any

import libsql

from database_mcp_core import Secret

_JS_SAFE_MAX = 2**53 - 1


def _json_safe(value: Any) -> Any:
    """Integers beyond the JS safe range are returned as strings (JSON-safe)."""
    if isinstance(value, int) and not isinstance(value, bool) and abs(value) > _JS_SAFE_MAX:
        return str(value)
    return value


def _dict_rows(cur) -> list[dict[str, Any]]:
    cols = [d[0] for d in cur.description]
    return [dict(zip(cols, r)) for r in cur.fetchall()]


class LibsqlAdapter:
    engine = "libsql"

    def __init__(self, url: str, auth_token: Secret | None = None) -> None:
        self._url = url
        self._token = auth_token
        self._db = None
        self._pool: concurrent.futures.ThreadPoolExecutor | None = None

    def connect(self, *, read_only: bool) -> None:
        # The driver takes a plain path for local files; strip the file:
        # scheme the conformance harness (and the TS line) uses for them.
        url = self._url[5:] if self._url.startswith("file:") else self._url
        self._db = libsql.connect(url, auth_token=self._token.reveal() if self._token else "")
        if read_only:
            try:
                self._db.execute("PRAGMA query_only = ON")
            except Exception:
                # Remote servers may not honor per-session pragmas. The SQL
                # guard still blocks writes; for real protection use a
                # read-only token.
                print(
                    "libsql: session read-only unavailable on this server, use a read-only auth token",
                    file=sys.stderr,
                )
        # Fail fast on bad URL/credentials.
        self._db.execute("SELECT 1")
        self._pool = concurrent.futures.ThreadPoolExecutor(max_workers=1)

    def close(self) -> None:
        if self._pool is not None:
            self._pool.shutdown(wait=False)
        if self._db is not None:
            self._db.close()

    def query(self, sql: str, *, max_rows: int, timeout_ms: int) -> dict[str, Any]:
        assert self._db is not None and self._pool is not None
        # ponytail: on timeout the call is abandoned, not cancelled (the
        # worker thread keeps running), same semantics as the TS adapter's
        # Promise.race.
        try:
            cur = self._pool.submit(self._db.execute, sql).result(timeout=timeout_ms / 1000)
        except TimeoutError:
            raise RuntimeError(f"timeout: query exceeded {timeout_ms}ms") from None
        # Writes come back with an empty description (not None like stdlib sqlite3).
        if not cur.description:
            self._db.commit()
            return {"columns": [], "rows": [], "rowCount": max(cur.rowcount, 0), "truncated": False}
        columns = [d[0] for d in cur.description]
        fetched = cur.fetchmany(max_rows + 1)
        truncated = len(fetched) > max_rows
        rows = [{c: _json_safe(v) for c, v in zip(columns, r)} for r in fetched[:max_rows]]
        return {"columns": columns, "rows": rows, "rowCount": len(rows), "truncated": truncated}

    def list_tables(self) -> list[dict[str, Any]]:
        assert self._db is not None
        names = [
            r[0]
            for r in self._db.execute(
                "SELECT name FROM sqlite_master WHERE type = 'table' AND name NOT LIKE 'sqlite_%' ORDER BY name"
            ).fetchall()
        ]
        out = []
        for name in names:
            quoted = name.replace('"', '""')
            (n,) = self._db.execute(f'SELECT count(*) FROM "{quoted}"').fetchone()
            out.append({"name": name, "estimatedRows": n})
        return out

    def describe_table(self, table: str) -> dict[str, Any]:
        assert self._db is not None
        cols = _dict_rows(self._db.execute("SELECT * FROM pragma_table_info(?)", (table,)))
        if not cols:
            raise LookupError(f"unknown table: {table}")

        columns = [
            {
                "name": c["name"],
                "type": c["type"],
                "nullable": not c["notnull"],
                "key": "PRI" if c["pk"] else None,
                "default": None if c["dflt_value"] is None else str(c["dflt_value"]),
            }
            for c in cols
        ]

        indexes = [
            {
                "name": ix["name"],
                "columns": [
                    r[0]
                    for r in self._db.execute(
                        "SELECT name FROM pragma_index_info(?)", (ix["name"],)
                    ).fetchall()
                ],
                "unique": bool(ix["unique"]),
            }
            for ix in _dict_rows(self._db.execute("SELECT * FROM pragma_index_list(?)", (table,)))
        ]

        fk_by_id: dict[int, dict[str, Any]] = {}
        for fk in _dict_rows(self._db.execute("SELECT * FROM pragma_foreign_key_list(?)", (table,))):
            entry = fk_by_id.setdefault(
                fk["id"],
                {
                    # SQLite-family FKs are unnamed; synthesize a stable one.
                    "name": f"fk_{table}_{fk['id']}",
                    "columns": [],
                    "referencesTable": fk["table"],
                    "referencesColumns": [],
                },
            )
            entry["columns"].append(fk["from"])
            entry["referencesColumns"].append(fk["to"])

        return {"name": table, "columns": columns, "indexes": indexes, "foreignKeys": list(fk_by_id.values())}
